rllib/agents/dqn/apex.py

```diff
@@ -283,6 +374,256 @@ def add_apex_metrics(result: dict) -> dict:
+        if not self.workers.remote_workers():
+            with self._timers[SAMPLE_TIMER]:
+                local_sampling_worker = self.workers.local_worker()
+                batch = local_sampling_worker.sample()
+                actor = random.choice(self.replay_actors)
+                ray.get(actor.add_batch.remote(batch))
+                batch_statistics = {
+                    local_sampling_worker: [
+                        {
+                            "agent_steps": batch.agent_steps(),
+                            "env_steps": batch.env_steps(),
+                        }
+                    ]
+                }
+                return batch_statistics
+
+        def sample_and_store(worker: RolloutWorker, replay_actors: List[ReplayActor]):
+            # how is the replay actor seeded?
+            # maybe we should use numpy, and pass the numpy random state of the
+            # driver to this function? TODO Avnish Sven Jun
+            _batch = worker.sample()
+            _actor = random.choice(replay_actors)
+            _actor.add_batch.remote(_batch)
+            _batch_statistics = {
+                "agent_steps": _batch.agent_steps(),
+                "env_steps": _batch.env_steps(),
+            }
+            return _batch_statistics
+
+        # Sample and Store in the Replay Actors on the sampling workers.
+        with self._timers[SAMPLE_TIMER]:
+            # Results are a mapping from ActorHandle (RolloutWorker) to their
+            # returned gradient calculation results.
+            num_samples_ready_dict: Dict[
+                ActorHandle, T
+            ] = asynchronous_parallel_requests(
+                remote_requests_in_flight=self.remote_sampling_requests_in_flight,
+                actors=self.workers.remote_workers(),
+                ray_wait_timeout_s=0.0,
+                max_remote_requests_in_flight_per_actor=4,
+                remote_fn=sample_and_store,
+                remote_kwargs=[{"replay_actors": self.replay_actors}]
+                * len(self.workers.remote_workers()),
+            )
+        return num_samples_ready_dict
+
+    def update_workers(self, _num_samples_ready: Dict[ActorHandle, int]) -> int:
+        """Update the remote workers that have samples ready.
+
+        Args:
+            _num_samples_ready: A mapping from ActorHandle (RolloutWorker) to
+                the number of samples returned by the remote worker.
+        Returns:
+            The number of remote workers whose weights were updated.
+        """
+        max_steps_weight_sync_delay = self.config["optimizer"]["max_weight_sync_delay"]
+        # Update our local copy of the weights if the learner thread has updated
+        # the learner worker's weights
+        if self.learner_thread.weights_updated:
+            self.learner_thread.weights_updated = False
+            weights = self.workers.local_worker().get_weights()
+            self.curr_learner_weights = ray.put(weights)
```

Comment:
should quickly return if weights have not been updated?

Reply:
hmmm no -- what this if statement does is places the learner's weights inside of the object store, but we still need to run the rest of the function, which triggers an update operation on the weights of the remote sampling workers.